File: python/packages/autogen-core/src/autogen_core/_cancellation_token.py
import asyncio
import threading
import time
from asyncio import Future
from typing import Any, Callable, List, Optional


class CancellationToken:
    """A token used to cancel pending async calls"""

    def __init__(self) -> None:
        self._cancelled: bool = False
        self._lock: threading.Lock = threading.Lock()
        self._callbacks: List[Callable[[], None]] = []

    def cancel(self) -> None:
        """Cancel pending async calls linked to this cancellation token."""
        with self._lock:
            if not self._cancelled:
                self._cancelled = True
                for callback in self._callbacks:
                    callback()

    def is_cancelled(self) -> bool:
        """Check if the CancellationToken has been used"""
        with self._lock:
            return self._cancelled

    def add_callback(self, callback: Callable[[], None]) -> None:
        """Attach a callback that will be called when cancel is invoked"""
        with self._lock:
            if self._cancelled:
                callback()
            else:
                self._callbacks.append(callback)

    def link_future(self, future: Future[Any]) -> Future[Any]:
        """Link a pending async call to a token to allow its cancellation"""
        with self._lock:
            if self._cancelled:
                future.cancel()
            else:

                def _cancel() -> None:
                    future.cancel()

                self._callbacks.append(_cancel)
        return future

    def child(self) -> "CancellationToken":
        """Create a child token that will be cancelled when this token is cancelled"""
        child = CancellationToken()

        def _cancel_child() -> None:
            child.cancel()

        self.add_callback(_cancel_child)
        return child

    @classmethod
    def combine(cls, *tokens: "CancellationToken") -> "CancellationToken":
        """Combine multiple tokens - the result will be cancelled when any of the input tokens is cancelled"""
        combined = cls()

        def _cancel_combined() -> None:
            combined.cancel()

        for token in tokens:
            token.add_callback(_cancel_combined)

        return combined

    @classmethod
    def with_timeout(cls, timeout_seconds: float) -> "CancellationToken":
        """Create a token that will be cancelled after a timeout"""
        token = cls()

        async def _timeout_task() -> None:
            await asyncio.sleep(timeout_seconds)
            token.cancel()

        # Schedule the timeout task
        try:
            loop = asyncio.get_running_loop()
            loop.create_task(_timeout_task())
        except RuntimeError:
            # No event loop running, use threading timer as fallback
            def _timeout_callback() -> None:
                token.cancel()

            timer = threading.Timer(timeout_seconds, _timeout_callback)
            timer.start()

        return token

    async def cancelled(self) -> None:
        """Wait for cancellation asynchronously"""
        if self.is_cancelled():
            return

        # Create a future that will be resolved when cancelled
        future: Future[None] = asyncio.Future()

        def _on_cancel() -> None:
            if not future.done():
                future.set_result(None)

        self.add_callback(_on_cancel)

        # If already cancelled, resolve immediately
        if self.is_cancelled() and not future.done():
            future.set_result(None)

        await future

    def check_cancelled(self) -> None:
        """Check if cancelled and raise an exception if so"""
        if self.is_cancelled():
            raise asyncio.CancelledError("Operation was cancelled")

File: rust/autogen-core/src/cancellation_token.rs
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::fmt;
use std::sync::{Arc, Mutex};
use tokio::sync::Notify;
use thiserror::Error;

/// Errors that can occur with cancellation tokens
#[derive(Error, Debug)]
pub enum CancellationError {
    #[error("Operation was cancelled")]
    Cancelled,
    #[error("Timeout occurred")]
    Timeout,
}

type Callback = Box<dyn Fn() + Send + Sync>;

#[derive(Clone)]
pub struct CancellationToken {
    inner: Arc<Mutex<Inner>>,
    notify: Arc<Notify>,
}

#[derive(Default)]
struct Inner {
    cancelled: bool,
    callbacks: Vec<Callback>,
}

impl CancellationToken {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(Inner::default())),
            notify: Arc::new(Notify::new()),
        }
    }

    pub fn cancel(&self) {
        let mut inner = self.inner.lock().unwrap();
        if !inner.cancelled {
            inner.cancelled = true;
            for callback in &inner.callbacks {
                callback();
            }
            inner.callbacks.clear();
            // Notify all waiting tasks
            self.notify.notify_waiters();
        }
    }

    pub fn is_cancelled(&self) -> bool {
        self.inner.lock().unwrap().cancelled
    }

    pub fn add_callback<F>(&self, callback: F)
    where
        F: Fn() + Send + Sync + 'static,
    {
        let mut inner = self.inner.lock().unwrap();
        if inner.cancelled {
            callback();
        } else {
            inner.callbacks.push(Box::new(callback));
        }
    }

    /// Wait for cancellation asynchronously
    pub async fn cancelled(&self) {
        if self.is_cancelled() {
            return;
        }
        self.notify.notified().await;
    }

    /// Check if cancelled and return an error if so
    pub fn check_cancelled(&self) -> Result<(), CancellationError> {
        if self.is_cancelled() {
            Err(CancellationError::Cancelled)
        } else {
            Ok(())
        }
    }

    /// Create a child token that will be cancelled when this token is cancelled
    pub fn child(&self) -> CancellationToken {
        let child = CancellationToken::new();
        let child_clone = child.clone();
        self.add_callback(move || {
            child_clone.cancel();
        });
        child
    }

    /// Combine multiple tokens - the result will be cancelled when any of the input tokens is cancelled
    pub fn combine(tokens: Vec<CancellationToken>) -> CancellationToken {
        let combined = CancellationToken::new();
        for token in tokens {
            let combined_clone = combined.clone();
            token.add_callback(move || {
                combined_clone.cancel();
            });
        }
        combined
    }

    /// Create a token that will be cancelled after a timeout
    pub fn with_timeout(duration: std::time::Duration) -> CancellationToken {
        let token = CancellationToken::new();
        let token_clone = token.clone();
        tokio::spawn(async move {
            tokio::time::sleep(duration).await;
            token_clone.cancel();
        });
        token
    }
}

impl fmt::Debug for CancellationToken {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("CancellationToken")
            .field("cancelled", &self.is_cancelled())
            .finish()
    }
}

impl Serialize for CancellationToken {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_bool(self.is_cancelled())
    }
}

impl<'de> Deserialize<'de> for CancellationToken {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let cancelled = bool::deserialize(deserializer)?;
        Ok(CancellationToken {
            inner: Arc::new(Mutex::new(Inner {
                cancelled,
                callbacks: Vec::new(),
            })),
            notify: Arc::new(Notify::new()),
        })
    }
}

impl Default for CancellationToken {
    fn default() -> Self {
        Self {
            inner: Arc::new(Mutex::new(Inner::default())),
            notify: Arc::new(Notify::new()),
        }
    }
}